flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/ChangelogRegistryImpl.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.changelog.fs;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.runtime.state.PhysicalStateHandleID;
+import org.apache.flink.runtime.state.StreamStateHandle;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.util.Map;
+import java.util.Set;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.CopyOnWriteArraySet;
+import java.util.concurrent.Executor;
+
+@Internal
+@ThreadSafe
+class ChangelogRegistryImpl implements ChangelogRegistry {
+    private static final Logger LOG = LoggerFactory.getLogger(ChangelogRegistryImpl.class);
+
+    private final Map<PhysicalStateHandleID, Set<UUID>> entries = new ConcurrentHashMap<>();
+    private final Executor executor;
+
+    public ChangelogRegistryImpl(Executor executor) {
+        this.executor = executor;
+    }
+
+    @Override
+    public void startTracking(StreamStateHandle handle, Set<UUID> backendIDs) {
+        LOG.debug(
+                "start tracking state, key: {}, state: {}",
+                handle.getStreamStateHandleID(),
+                handle);
+        entries.put(handle.getStreamStateHandleID(), new CopyOnWriteArraySet<>(backendIDs));
+    }
+
+    @Override
+    public void stopTracking(StreamStateHandle handle) {
+        LOG.debug(
+                "stop tracking state, key: {}, state: {}", handle.getStreamStateHandleID(), handle);
+        entries.remove(handle.getStreamStateHandleID());
+    }
+
+    @Override
+    public void notUsed(StreamStateHandle handle, UUID backendId) {
```

Comment:
The notification of checkpoint complete cannot alaways be guranteed, is it always safe here to call this method considering `stopTracking` missed?

Reply:
Good question.

It is the client responsibility to call `notUsed` when it is actually safe to do so. This is achieved by calling `notUsed` on checkpoint **subsumption** notification. And when a checkpoint is subsumed, notification about its **completion** doesn't matter.

And when there are several backends involved, it is the responsibility of the Registry to wait for `notUsed()` call from **all** backends before discarding the state (and not discard it if at least one called `stopTracking()`).

Reply:
If current sequense number is up to `100`, we then call `changelogTruncateHelper.materialized(upTo)` after materialization finished. Thus, we would discard change logs with `upTo` as `100`, however, if the `stopTracking` is missed before, can we say the discarding is safe here? Considering the next checkpoints are not successful and thus the materialized parts are not be used.

Reply:
> we then call changelogTruncateHelper.materialized(upTo) after materialization finished. 
> Thus, we would discard change logs with upTo as 100

Let's assume there is some valid or pending checkpoint that uses sqn <= 100. But JM completion notification was lost, and therefore `stopTracking` was not called.

But if that checkpoint is valid/pending, it couldn't be subsumed (nor any newer checkpoint).
Therefore, `ChangelogTruncateHelper.subsumedUpTo` hasn't been advanced yet.
Therefore, discard will not happen at this point because `ChangelogTruncateHelper` truncates the logs up to `min(materializedUpTo, subsumedUpTo)`.

Reply:
I see, and I think this is valid. However, this guarantee seems not tested in `ChangelogStateDiscardTest`.

Reply:
Isn't this covered in `testPreEmptiveUploadDiscardedOnSubsumption`:
```
materialize(backend, writer);
checkpoint(backend, 1L);
assertRetained(uploader.results); // may be used in non-subsumed checkpoints
```
?

Reply:
I think the test below should also be verified:
~~~java
changeAndLogRandomState(backend, uploader.results::size);
checkpoint(backend, 1L);
materialize(backend, writer);
assertRetained(uploader.results); // may be used in non-subsumed checkpoints
~~~

Even we call the `materialize` after `checkpoint`, since we did not receive checkpoint complete message, the uploader results are still there.

Reply:
Added `testPreEmptiveUploadNotDiscardedWithoutNotification` in 5ea7f3650ee.